File: 02/017/main.py
import re

def getBritishVocal(num):
    nums = {
        0:  '',
        1:  'one',
        2:  'two',
        3:  'three',
        4:  'four',
        5:  'five',
        6:  'six',
        7:  'seven',
        8:  'eight',
        9:  'nine',
        10:  'ten',
        11:  'eleven',
        12:  'twelve',
        13:  'thirteen',
        14:  'fourteen',
        15:  'fifteen',
        16:  'sixteen',
        17:  'seventeen',
        18:  'eighteen',
        19:  'nineteen',
        20:  'twenty',
        30:  'thirty',
        40:  'forty',
        50:  'fifty',
        60:  'sixty',
        70:  'seventy',
        80:  'eighty',
        90:  'ninety',
    }
    if nums.get(num) != None:
        return nums.get(num)
    if num > 999:
        thousandth = int(num / 1000)
        rest = num - 1000 * thousandth
        vocalThousandth = getBritishVocal(thousandth) + ' thousand'
        vocalRest = ' and ' + getBritishVocal(rest) if rest > 0 else ''
        return vocalThousandth + vocalRest
    if num > 99:
        hundredth = int(num / 100)
        rest = num - 100 * hundredth
        vocalHundred = getBritishVocal(hundredth) + ' hundred'
        vocalRest = ' and ' + getBritishVocal(rest) if rest > 0 else ''
        return vocalHundred + vocalRest
    if num > 20:
        tenth = int(num / 10) * 10
        onceth = num % 10
        return getBritishVocal(tenth) + '-' + getBritishVocal(onceth)

def numberLetter():
    return sum([len(re.sub('[ -]', '', a)) for a in [getBritishVocal(b) for b in range(1, 1001)]])

print(numberLetter())

File: 02/017/main.hs
getBritishVocal 0 = ""
getBritishVocal 1 = "one"
getBritishVocal 2 = "two"
getBritishVocal 3 = "three"
getBritishVocal 4 = "four"
getBritishVocal 5 = "five"
getBritishVocal 6 = "six"
getBritishVocal 7 = "seven"
getBritishVocal 8 = "eight"
getBritishVocal 9 = "nine"
getBritishVocal 10 = "ten"
getBritishVocal 11 = "eleven"
getBritishVocal 12 = "twelve"
getBritishVocal 13 = "thirteen"
getBritishVocal 14 = "fourteen"
getBritishVocal 15 = "fifteen"
getBritishVocal 16 = "sixteen"
getBritishVocal 17 = "seventeen"
getBritishVocal 18 = "eighteen"
getBritishVocal 19 = "nineteen"
getBritishVocal 20 = "twenty"
getBritishVocal 30 = "thirty"
getBritishVocal 40 = "forty"
getBritishVocal 50 = "fifty"
getBritishVocal 60 = "sixty"
getBritishVocal 70 = "seventy"
getBritishVocal 80 = "eighty"
getBritishVocal 90 = "ninety"
getBritishVocal num
    | num > 999     = getBritishVocal thousandth ++ " thousand" ++ vocalRestThousand
    | num > 99      = getBritishVocal hundredth ++ " hundred" ++ vocalRestHundred
    | num > 20      = getBritishVocal (div num 10 * 10) ++ "-" ++ getBritishVocal(mod num 10)
    where
        thousandth = div num 1000
        hundredth = div num 100
        restThousand = num - 1000 * thousandth
        restHundred = num - 100 * hundredth
        vocalRestThousand = if restThousand > 0 then " and " ++ getBritishVocal restThousand else ""
        vocalRestHundred = if restHundred > 0 then " and " ++ getBritishVocal restHundred else ""

numberVocalCount num = length [a | a <- getBritishVocal num, a /= ' ' && a /= '-']

numberLetterCounts = sum [numberVocalCount a | a <- [1..1000]]

main = do
    print numberLetterCounts
